DataFormats/TrivialSerialisation/interface/MemoryCopyTraits.h

```diff
@@ -0,0 +1,145 @@
+  // source object to a destination object.
+  //
+  //
+  // A specialisation may implement the method properties(), which returns the
+  // properties of an existing object, which can be used to initialize a newly
+  // allocated copy of the object via the initialize() method.
+  //
+  //   using Properties = ...;
+  //   static Properties properties(T const& object);
+  //
+  // If properties() is not implemented, the initialize() method takes a single
+  // argument:
+  //
+  //   static void initialize(T& object);
+  //
+  // If properties() is implemented, the initialize() method should take as a
+  // second parameter a const reference to a Properties object:
+  //
+  //   static void initialize(T& object, Properties const& args);
+  //
+  //
+  // A specialisation can optionally provide a static method
+  //
+  //   static void finalize(T& object);
+  //
+  // If present, it should be called to restore the object invariants after a
+  // memcpy operation.
+  //
+
+  template <typename T>
+  struct MemoryCopyTraits;
+
+  // Checks if the properties method is defined
+  template <typename T>
+  concept HasTrivialCopyProperties = requires(T const& object) { MemoryCopyTraits<T>::properties(object); };
+
+  // Get the return type of properties(...), if it exists.
+  template <typename T>
+    requires HasTrivialCopyProperties<T>
+  using TrivialCopyProperties = decltype(MemoryCopyTraits<T>::properties(std::declval<T const&>()));
+
+  // Checks if the declaration of initialize(...) is consistent with the presence or absence of properties.
+  template <typename T>
+  concept HasValidInitialize =
+      // does not have properties(...) and initialize(object) takes a single argument, or
+      (not HasTrivialCopyProperties<T> and requires(T& object) { MemoryCopyTraits<T>::initialize(object); }) or
+      // does have properties(...) and initialize(object, props) takes two arguments
+      (HasTrivialCopyProperties<T> and
+       requires(T& object, TrivialCopyProperties<T> props) { MemoryCopyTraits<T>::initialize(object, props); });
+
+  // Checks for const and non const memory regions
+  template <typename T>
+  concept HasRegions = requires(T& object, T const& const_object) {
+    MemoryCopyTraits<T>::regions(object);
+    MemoryCopyTraits<T>::regions(const_object);
+  };
+
+  // Checks if there is a valid specialisation of MemoryCopyTraits for a type T
+  template <typename T>
+  concept HasMemoryCopyTraits =
+      // Has memory regions declared and
+      (HasRegions<T>) and
+      // has either no initialize(...) or a valid one
+      (not requires { &MemoryCopyTraits<T>::initialize; } or HasValidInitialize<T>);
+
+  // Checks if finalize(...) is defined
+  template <typename T>
+  concept HasTrivialCopyFinalize = requires(T& object) { ngt::MemoryCopyTraits<T>::finalize(object); };
+
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+  // Specialisations for various types
```

Comment:
Do these specializations need to be visible to everyone who `#include`s this header, or only to the plugins where the `DEFINE_TRIVIAL_SERIALISER_PLUGIN` is used to define the plugins for the specialized types?

Reply:
In principle the `MemoryCopyTraits` are more generic than the trivial serialisation method - for example, they should provide enough information to reimplement the host/device copies, if we wanted to.

For this reason I would prefer not to move the specialisation to the plugin that uses them.

I can move them to separate headers in the same package, _e.g._ `DataFormats/TrivialSerialisation/interface/arithmetic.h`, `DataFormats/TrivialSerialisation/interface/vector.h`, _etc_ if you think that would be clearer.
but I'm not sure if that would be better ?